independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    /**
+     * This buffer append data in multiples {@link ByteBuf}s which minimum capacity is {@code minChunkSize} or
+     * as each {@code len}, if greater than it.<br>
+     * The data is consolidated in a single {@link CompositeByteBuf} on {@link #clear}.
+     */
+    public static AppendBuffer withMinChunks(ByteBufAllocator allocator, int capacity, int minChunkSize) {
+        return new AppendBuffer(allocator, capacity, minChunkSize);
+    }
+
+    private ByteBuf lastBuffer() {
+        if (otherBuffers == null || otherBuffers.isEmpty()) {
+            return buffer;
+        }
+        return otherBuffers.peekLast();
+    }
+
+    /**
+     * It returns how many bytes have been appended<br>
+     * If returns a value different from {@code len}, is it required to invoke {@link #clear}
+     * that would refill the available capacity till {@link #capacity()}
+     */
+    public int append(byte[] bytes, int off, int len) {
+        int alreadyWritten = 0;
+        if (minChunkSize > 0) {
+            var lastBuffer = lastBuffer();
+            if (lastBuffer != null) {
+                int availableOnLast = lastBuffer.writableBytes();
+                if (availableOnLast > 0) {
+                    int toWrite = Math.min(len, availableOnLast);
+                    lastBuffer.writeBytes(bytes, off, toWrite);
+                    size += toWrite;
+                    len -= toWrite;
+                    // we stop if there's no more to append
+                    if (len == 0) {
+                        return toWrite;
+                    }
+                    off += toWrite;
+                    alreadyWritten = toWrite;
+                }
+            }
+        }
+        final int availableCapacity = capacity - size;
+        if (availableCapacity == 0) {
+            return alreadyWritten;
+        }
+        // we can still write some
+        int toWrite = Math.min(len, availableCapacity);
+        final int chunkCapacity;
+        if (minChunkSize > 0) {
+            // Cannot allocate less than minChunkSize, till the limit of capacity left
+            chunkCapacity = Math.min(Math.max(minChunkSize, toWrite), availableCapacity);
+        } else {
+            chunkCapacity = toWrite;
+        }
+        var tmpBuf = allocator.directBuffer(chunkCapacity);
+        tmpBuf.writeBytes(bytes, off, toWrite);
+        if (buffer == null) {
+            buffer = tmpBuf;
+        } else {
+            boolean newQ = false;
+            try {
+                if (otherBuffers == null) {
+                    otherBuffers = new ArrayDeque<>();
+                    newQ = true;
+                }
+                otherBuffers.add(tmpBuf);
+            } catch (Throwable t) {
+                // rollback to preserve state
+                if (newQ) {
+                    otherBuffers = null;
+                }
+                size -= alreadyWritten;
+                tmpBuf.release();
```

Comment:
I didn't rollback the lastBuffer's writer Index either here (in case it has been written) :(